numpy/lib/_datasource.py

```diff
@@ -38,13 +38,59 @@
 import os
 import sys
 import shutil
+import io
 
 _open = open
 
+def _check_mode(mode, encoding, newline):
+    if "t" in mode:
+        if "b" in mode:
+            raise ValueError("Invalid mode: %r" % (mode,))
+    else:
+        if encoding is not None:
+            raise ValueError("Argument 'encoding' not supported in binary mode")
+        if newline is not None:
+            raise ValueError("Argument 'newline' not supported in binary mode")
+
+def _python2_bz2open(fn, mode, encoding, newline):
+    """ wrapper to open bz2 in text mode """
```

Comment:
Needs docstring of the standard type with expanded explanation and documentation of parameters.